### ISSUE TITLE
systemd-resolved: Include `ad` flag and EDNS/EDE data without doing DNSSEC validation

### ISSUE THREAD

**Issue by Anuskuss** [systemd/systemd#31365]
### Component

systemd-resolved

### Is your feature request related to a problem? Please describe

My upstream DNS already does DNSSEC validation and because I'm doing DoT (I trust the server and the data after it arrives) I'm confident that everything is in order. That's why I'm suggesting the ability to just forward the DNSSEC data, in particular the `ad` flag on success

```
$ dig @1.1.1.1 sigok.ippacket.stream
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 24296
;; flags: qr rd ra ad; QUERY: 1, ANSWER: 2, AUTHORITY: 0, ADDITIONAL: 1
```

and the EDE data on failure

```
$ dig @1.1.1.1 sigfail.ippacket.stream
;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 1232
; EDE: 6 (DNSSEC Bogus): (failed to verify sigfail.rsa2048-sha256.ippacket.stream. A: using DNSKEY ids = [46436])
```

### Describe the solution you'd like

DNSSEC: `bool | "allow-downgrade" | "proxy"`

### Describe alternatives you've considered

Always include the data when DNSSEC is disabled or inactive.

### The systemd version you checked that didn't have the feature you are asking for

255

**Comment by Anuskuss** [comment]
So does `allow-downgrade` though and it's the reason it's not the default - it's a deliberate choice a user has to make. And it's free DNSSEC if you trust your upstream (for me personally, DNSSEC is pointless and it's not worth doing extra work for it).

**Comment by letoams** [comment]
The AD flag’s meaning is defined by https://datatracker.ietf.org/doc/html/rfc4035#section-3.2.3

You cannot change its meaning because then consumers of systemd-resolved can’t trust the AD flag anymore (and ironically, would have to do their own validation on top)

The AD flag means 1 thing and you are asking for that one thing to be removed from the flag. Why aren’t you just going to ignore the presence of the flag ?

**Comment by Anuskuss** [comment]
> Why aren’t you just going to ignore the presence of the flag ?

I don't understand what you mean. I'm proposing the ability for systemd-resolved to *pass on* the `ad` flag, meaning set it if the upstream server also sets it. Not set it for every request (i.e. non-successful DNSSEC).

**Comment by letoams** [comment]
That is not what the AD flag means:

The name server side of a security-aware recursive name server MUST
   NOT set the AD bit in a response unless the name server considers all
   RRsets in the Answer and Authority sections of the response to be
   authentic.  The name server side SHOULD set the AD bit if and only if
   the resolver side considers all RRsets in the Answer section and any
   relevant negative response RRs in the Authority section to be
   authentic.  The resolver side MUST follow the procedure described in
   [Section 5](https://datatracker.ietf.org/doc/html/rfc4035#section-5) to determine whether the RRs in question are authentic.
   However, for backward compatibility, a recursive name server MAY set
   the AD bit when a response includes unsigned CNAME RRs if those CNAME [RFC 4035]   RRs demonstrably could have been synthesized from an authentic DNAME
   RR that is also included in the response according to the synthesis
   rules described in [[RFC2672](https://datatracker.ietf.org/doc/html/rfc2672)].



If you are not validating, you cannot set the AD bit and must not relay any AD bits.

**Comment by Anuskuss** [comment]
> If you are not validating, you cannot set the AD bit and must not relay any AD bits.

Yeah, I understand that part. I want to violate the spec though because like I said, I find DNSSEC to be pretty pointless (less than 1% of my queries can be successfully validated) and wouldn't want to waste CPU cycles doing it. And relaying the AD bit is free.

Should I split this FR into two then?  `DNSSEC: proxy` may be too controversial but the second part sounds like a no-brainer. Or do you have some comments about that as well?

**Comment by letoams** [comment]
Your use case is the few operations of DNSSEC compared to like your whole disk encrypted LUKS crypto ? Against all TLS? All the JavaScript in all your browser tabs ? 

anyway, DNSSEC=off let’s you disable DNSSEC.

**Comment by TimStallard** [comment]
This would be useful where a DNSSEC validating resolver is in use, but modifies the answers after validation, for example with DNS64. If applications require visibility of the ad flag, currently resolved can't be used in this setup - with DNSSEC=off the ad flag will never be set, but with resolved's DNSSEC validation enabled the modified answers will fail validation and result in a servfail.

**Comment by letoams** [comment]
On Mon, 19 Feb 2024, Tim Stallard wrote:

> This would be useful where a DNSSEC validating resolver is in use, but modifies the answers after validation, for example
> with DNS64.

DNS64 with NAT64 might be the only case where this is okay, but it means
it would still be resolving and validating the DNS replies from the
network and only then locally synthesize the answer. So it has confirmed
the DNSSEC validation of all data received over the network. That wasn't
the case requested here, where the AD bit is received from upstream
and relayed.

> If applications require visibility of the ad flag, currently resolved can't be used in this setup - with
> DNSSEC=off the ad flag will never be set, but with resolved's DNSSEC validation enabled the modified answers will fail
> validation and result in a servfail.

Sadly, this is not true. There are many ways in which resolved butchers
things and lies to the local client, eg see:

https://github.com/systemd/systemd/issues/25676

where Lennart wrote:

 	The DNSSEC support in resolved is off by default for a reason, and it's
 	not complete. It's very hard to get this work in a reasonable way given
 	the state of DNSSEC servers in the wild. Before we can actually recommend
 	people to use this we have to tweak this code a lot still. DNSSEC on
 	the client is on the backburner, since it's such a massive amount of
 	work to make it reasonably work with the end-user routers out there.

 	I am happy to review patches, but just bitching about an experimental
 	feature all the time really doesn't help anyone. It just makes me lower
 	my interest in dealing with these issues.

ExecSum: systemd-resolved is not RFC compliant, contains CVEs closed as
"documentation given" and will not be fixed in the foreseable future. It
has been two weeks since the reports of CVE-2023-50387 and CVE-2023-50868
and systemd-resolved is the only DNS resolver software that has not made
a statement it is not vulnerable nor delivered patches or a new version.

Go ahead and muck with the AD flag. More RFC violations at this point
hardly matters. systemd-resolved is abandonware and does not implement
the core DNS RFCs.

**Comment by Anuskuss** [comment]
Not that anyone asked but I ran some tests on a few DNS proxies I found that support DoT:
Service|μ ms|σ ms|q/s|✓|ad|EDE
--:|:-:|:-:|:-:|:-:|:-:|:-:
[systemd-resolved](https://github.com/systemd/systemd)|148|339|631|99.80%||
[coredns](https://github.com/coredns/coredns)|150|324|565|99.46%|x|x
[routedns](https://github.com/folbricht/routedns)|150|360|619|99.77%|x|x
[dns-over-tls-forwarder](https://github.com/mikispag/dns-over-tls-forwarder)|152|362|604|99.73%|x|
[dnsdist](https://github.com/PowerDNS/pdns)|156|365|598|99.79%|x|x
[dnsproxy](https://github.com/AdguardTeam/dnsproxy)|163|364|576|99.79%|x|
[stubby](https://github.com/getdnsapi/stubby)|165|353|568|99.79%|x|

`systemd-resolved` is the only one out there **not** doing what I want it to do ("violate the spec") **but** is also the fastest. `routedns` is the 3rd fastest (within a ms of 2nd) and seems to be comparable to `systemd-resolved` (takes 2 ms longer and loses an additional 221 queries out of 1M). It's really a toss-up between raw speed and features.

> systemd-resolved is abandonware and does not implement the core DNS RFCs.

I don't know who you are and if that's just your opinion or the general consensus amongst systemd developers but if that's really the case, I think I'll switch to `routedns` for now. I really don't want to have to read about a vulnerability in the future just because nobody maintains systemd's DNS code anymore. That and I don't like the `.service` file (you can't override dependencies and `systemd-resolved` unfortunately defines `Before=network.target`).

**Comment by paulwouters** [comment]
Note that I am not a systemd developer. I do however, write DNS RFCs at IETF, see https://datatracker.ietf.org/person/paul.wouters@aiven.io

I still believe the AD flag should be cleared by DNS proxies that do not validate themselves according to the RFCs.
If they wanted to optimize validating without too many RTTs, they could implement DNS Chain Query Requests of RFC7901  https://datatracker.ietf.org/doc/rfc7901/

Thanks for providing this data btw.

**Comment by paulwouters** [comment]
note: if you tested systemd-resolved in default config (eg with DNSSEC=no) that might explain the lack of AD flag. If you tested with DNSSEC=yes, then I'd be pleasantly surprised it did actually clear it, as it is the right thing to do!

**Comment by rpigott** [comment]
[RFC4035§4.9.3](https://datatracker.ietf.org/doc/html/rfc4035#section-4.9.3) quite explicitly allows this behavior:

> A **non-validating security-aware stub resolver** MAY chose to examine the setting of the AD bit in response messages that it receives in order to determine whether the security-aware recursive name server that sent the response claims to have cryptographically verified the data in the Answer and Authority sections of the response message.
> [...]
> a security-aware stub resolver MUST NOT place any reliance on signature validation allegedly performed on its behalf, **except when the security-aware stub resolver obtained the data in question from a trusted security-aware recursive name server via a secure channel.**

As sd-resolved is a stub resolver, this is the paragraph that applies.

**Comment by letoams** [comment]
On Fri, 1 Mar 2024, Ronan Pigott wrote:

> RFC4035§4.9.3 quite explicitly allows this behavior:
>
>       A non-validating security-aware stub resolver MAY chose to examine the setting of the
>       AD bit in response messages that it receives in order to determine whether the
>       security-aware recursive name server that sent the response claims to have
>       cryptographically verified the data in the Answer and Authority sections of the
>       response message.
>       [...]
>       a security-aware stub resolver MUST NOT place any reliance on signature validation
>       allegedly performed on its behalf, except when the security-aware stub resolver
>       obtained the data in question from a trusted security-aware recursive name server via
>       a secure channel.
> 
> As sd-resolved is a stub resolver, this is the paragraph that applies.

It's not a simple stub resolver. It is a recursive DNS / proxy combo.
It sends all queries over all interfaces and picks the fastests response
without guarantee, eg stubs using systemd-resolved do not know if the
query came over DoT or DoH or in the clear. They rely on the AD bit
to carry meaning.

glibc getaddrinfo() is a stub resolver. It does not do any recursive
DNS, caching, etc. It cannot do validation itself, it outsources that
to the recursive resolver it is using - eg like systemd-resolved.

systemd-resolved is a full validating resolver.  You expect it to
validate and not proxy AD bits around. If it just relays DNS answers
without doing the validation itself, it should not set the AD bit.

**Comment by rpigott** [comment]
No, it is a non-validating security-aware stub resolver (when operating with dnssec=no). The definition is available in a sibling rfc if you're curious: [RFC4033§2](https://datatracker.ietf.org/doc/html/rfc4033#section-2). And to save you time on the quiz, getaddrinfo is none of the above.

**Comment by mildsunrise** [comment]
> It's not a simple stub resolver. It is a recursive DNS / proxy combo.

systemd-resolved implements a stub resolver, not a recursive resolver. it has no knowledge of the root nameservers. it doesn't follow delegations or CNAMEs. upstream nameservers given to systemd-resolved must support recursive queries.

the difference between glibc's stub resolver and systemd-resolved is that glibc only exposes a POSIX interface (getaddrinfo and friends) while systemd-resolved also exposes other interfaces, among them a DNS interface (bound at 127.0.0.53:53).

(there's also the proxy (bound at 127.0.0.54:53) but that functionality would not be affected by the change being proposed here)

therefore I fully agree with this interpretation:

> it is a non-validating security-aware stub resolver (when operating with dnssec=no)

and I also request a `DNSSEC=trust-ad` setting, just like glibc's stub resolver's identically named one.
if glibc's `trust-ad` setting isn't a spec violation, then I don't think a `DNSSEC=trust-ad` mode would be either.

**Comment by nl6720** [comment]
This issue looks like a duplicate of https://github.com/systemd/systemd/issues/12317.
